Adm1266 program firmware

tachicialex opened the pull request:
This series adds the ability to write the firmware through PMBus by writing the Intel Hex file at the start of Nvmem file created by adm1266 driver:
- dt-bindings: hwmon: Add ref node for ADM1266 slaves: slaves now have a phandle to their master
- hwmon: (pmbus/core) Add Block W: used for writing the password as specified in AN-1453
- hwmon: (pmbus/core) Add group command support: send commands to all cascaded devices but execute them only when stop bit is sent
- hwmon: pmbus: adm1266: program firmware: writing to address 0 of nvmem file will now trigger a firmware update of all cascaded devices.

commodo commented:
commit comments seem to be [visibly] shorter than 75 columns;
you might get complaints upstream
if you add this to your ~/.vimrc file, this could work to help auto-break at 75 columns
`
autocmd BufRead,BufNewFile COMMIT_EDITMSG,*.patch setl textwidth=75
`

bonus: you can also add some spellcheck: `autocmd Filetype gitcommit setlocal spell`

commodo commented on drivers/hwmon/pmbus/pmbus_core.c:
```diff
@@ -318,6 +318,67 @@ int pmbus_block_write(struct i2c_client *client, u8 cmd, u8 w_len, u8 *data_w)
+		msgs[i].buf = write_buf,
+		msgs[i].len = w_lens[i] + 1,
+
+
```
nitpick: extra line

commodo commented on drivers/hwmon/pmbus/pmbus_core.c:
```diff
@@ -318,6 +318,67 @@ int pmbus_block_write(struct i2c_client *client, u8 cmd, u8 w_len, u8 *data_w)
+	ret = i2c_transfer(clients[0]->adapter, msgs, nr_cmds);
+
+cleanup:
+	for (i = i - 1; i >= 0; i--)
```
i think `i` can go below zero here, or?

tachicialex replied:
if i < 0 the code in loop won't get executed

commodo replied:
ack;

commodo commented on drivers/hwmon/pmbus/pmbus_core.c:
```diff
@@ -318,6 +318,67 @@ int pmbus_block_write(struct i2c_client *client, u8 cmd, u8 w_len, u8 *data_w)
+	int ret;
+	int i;
+
+	msgs = (struct i2c_msg *)kzalloc(nr_cmds * sizeof(struct i2c_msg),
```
the explicit cast is usually un-needed;

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -392,14 +472,361 @@ static int adm1266_nvmem_read(void *priv, unsigned int offset, void *val,
+	int ret;
+
+	fw = dev_fwnode(&data->client->dev);
+	INIT_LIST_HEAD(&data->cascaded_devices_list);
```
usually, you don't need to initialize the list entry; only the list-head;
when list_add() gets called on this [to add this entry to a list] all should be properly initialized;
otherwise, it's not being used;

commodo replied:
my bad; this is the list head;
need more coffee;

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -392,14 +472,361 @@ static int adm1266_nvmem_read(void *priv, unsigned int offset, void *val,
+			return -ENOMEM;
+
+		master_ref->data = data;
+		INIT_LIST_HEAD(&master_ref->list);
```
however, this can be removed; list_add() will initialize this properly and when added to the list;

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -392,14 +472,361 @@ static int adm1266_nvmem_read(void *priv, unsigned int offset, void *val,
+						 &master_fwnode_ref);
+	if (ret < 0) {
+		dev_err(&data->client->dev,
+			"Could not read adi,connected-adm1266 property");
```
`adi,connected-adm1266 ` -> `adi,master-adm1266 ` ?

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -392,14 +472,361 @@ static int adm1266_nvmem_read(void *priv, unsigned int offset, void *val,
+		return -ENOMEM;
+
+	slave_ref->data = slave;
+	INIT_LIST_HEAD(&slave_ref->list);
```
this can also be removed;

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -392,14 +472,361 @@ static int adm1266_nvmem_read(void *priv, unsigned int offset, void *val,
+	if (!fwnode_property_present(fw, "adi,master-adm1266")) {
+		data->master_dev = true;
+
+		master_ref = kzalloc(sizeof(*master_ref), GFP_KERNEL);
```
would `devm_kzalloc()` work?
then you don't need to implement any corresponding kfree() on error & exit paths

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -392,14 +472,361 @@ static int adm1266_nvmem_read(void *priv, unsigned int offset, void *val,
+{
+	struct adm1266_data_ref *slave_ref;
+
+	slave_ref = kzalloc(sizeof(*slave_ref), GFP_KERNEL);
```
same here: would `devm_kzalloc()` work?

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -437,6 +879,20 @@ static int adm1266_probe(struct i2c_client *client,
+		return ret;
+
+	if (!ret) {
+		dev_err(&client->dev, "Chip firmware not written.");
```
this doesn't look like an error; maybe `dev_info` or `dev_notice`?
i'm noticing, this doesn't initialize the debugfs;
would it be a good idea to init it?

we can leave that for later;
maybe just move `adm1266_debug_init()` before this call?

commodo replied:
naturally, the calls to the debugfs in this state would probably need to error out

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -392,14 +472,361 @@ static int adm1266_nvmem_read(void *priv, unsigned int offset, void *val,
+	return 0;
+}
+
+static int adm1266_register(struct adm1266_data *data)
```
there doesn't seem to be an equivalent unregister for this;
maybe now, you'd need to implement a custom remove hook that removes it from the master/cascaded list and calls `.pmbus_do_remove` after-wards;

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -392,14 +472,361 @@ static int adm1266_nvmem_read(void *priv, unsigned int offset, void *val,
+		return -EINVAL;
+	}
+
+	return 0;
```
remove this `return 0` ?

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -392,14 +472,361 @@ static int adm1266_nvmem_read(void *priv, unsigned int offset, void *val,
+						 ADM1266_FIRMWARE_SIZE);
+
+		if (fw_writen)
+			adm1266_program_firmware(data);
```
this looks like it could be `ret = adm1266_program_firmware()`

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -392,14 +472,361 @@ static int adm1266_nvmem_read(void *priv, unsigned int offset, void *val,
+	u8 *hex_cmd = data->dev_mem + offset;
+	u8 *fw_end = data->dev_mem + offset + size;
+
+	while (1) {
```
it's pretty hard to sell/convince these `while (1)` loops upstream;
can this be reworked to be somehow with a terminal condition?
reason is: when you look at this, you can an initial feeling this could do an endless loop;
so, even something like:
`while (!hex_cmd || hex_cmd >= fw_end)` or `do { } while (!hex_cmd || hex_cmd >= fw_end)` or even `for(; !hex_cmd || hex_cmd >= fw_end ;)` is quite nicer to convince upstream than `while(1)`

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -392,14 +472,361 @@ static int adm1266_nvmem_read(void *priv, unsigned int offset, void *val,
+	{ADM1266_STORE_USER_ALL, 300, 0},
+};
+
+static int adm1266_write_intel_hex(struct adm1266_data *data,
```
curios: why  `intel` ?
this could assume it's something about Intel [the company]

commodo replied:
maybe let's just call this `adm1266_write_hex()`

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -392,14 +472,361 @@ static int adm1266_nvmem_read(void *priv, unsigned int offset, void *val,
+	int ret;
+	int i;
+
+	memset(first_writes, 1, 7);
```
this sets `0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01`
did you want `0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff`?

commodo replied:
ah; 0x01 looks fine as well looking closer at the logic;

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -392,14 +472,361 @@ static int adm1266_nvmem_read(void *priv, unsigned int offset, void *val,
+
+	memset(first_writes, 1, 7);
+
+	while (1) {
```
also here;
what if we do something like `while (size > 0)`?

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -392,14 +472,361 @@ static int adm1266_nvmem_read(void *priv, unsigned int offset, void *val,
+
+		/* write to eeprom with specified delays */
+		write_delay = 40;
+		for (i = 0; i < 7; i++) {
```
this seems to be a quite a bit of logic for getting some write delays;
is the write delay important to be per-reg-address?
or is it important just the first time?

mixing the two like this, looks like can have some potential for weird timing bugs;
well, it could be fine as-is, but if we could pick one delay per reg-address, regardless of `first_writes`, it would be a bit easier to understand;
also, if the timing penalty [for not using the `first_writes` logic/variable] is not that great, i would not use it;

tachicialex replied:
At first i ignored the first_writes delays and i received a lot of random NACKs from the chip during the firmware/config programming. Waiting always the first delay is a huge time penalty leading to +10 min firmware programming.

The chip is very strict when it comes to delays.

commodo replied:
ack; thanks for the details;
the only thing to take care [when sending this upstream] is to have an explanation ready if there are questions; the code itself, is relatively easy to understand, there's only the case of motivating the "why this is needed like this" for upstream

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -243,6 +291,35 @@ static inline int adm1266_config_gpio(struct adm1266_data *data)
+	memset(w_lens, w_len, ADM1266_MAX_DEVICES);
+	memset(cmds, cmd, ADM1266_MAX_DEVICES);
+
+	if (to_slaves) {
```
maybe it wouldn't be too bad to just do:
```
if (!to_slaves)
    return pmbus_group_command(clients, cmds, w_lens, data_w, i);
```

and un-indent the rest

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -364,6 +441,9 @@ static int adm1266_read_mem_cell(struct adm1266_data *data,
 		return ret;
+	case ADM1266_FIRMWARE_OFFSET:
+		/* firmware is write-only */
+		return 0;
```
curios: should this be `return -EINVAL` or some error?
if it's usually acceptable to be `return 0` i'm also fine with this

tachicialex replied:
in case a user would like to read the whole nvmem file in one go.

commodo replied:
ack; thanks for the details;

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -392,14 +472,361 @@ static int adm1266_nvmem_read(void *priv, unsigned int offset, void *val,
+	}
+	if (ADM1266_STATUS_PART_LOCKED(reg_val)) {
+		dev_err(&client->dev, "Device locked.");
+		return -EINVAL;
```
usually -EBUSY is more suited for `part locked/busy`

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -392,14 +472,361 @@ static int adm1266_nvmem_read(void *priv, unsigned int offset, void *val,
+
+	for (i = 0; i < data->nvmem_config.ncells; i++) {
+		mem_cell = &adm1266_nvmem_cells[i];
+		if (adm1266_cell_is_accessed(mem_cell, offset, bytes)) {
```
you can reduce one level of indentation with:
```
if (!adm1266_write_intel_hex())
    continue;
```

commodo commented:
the series looks pretty good overall

tachicialex commented:
Changelog v1 -> v2:
- add check for null at `kzalloc` (replaced with `kcalloc`)
(in `pmbus_group_command`)
- now if `ADM1266_STATUS_PART_LOCKED` `return -EBUSY`
- rename `adm1266_write_intel_hex` to `adm1266_write_hex`
- in `adm1266_write_hex` changed `while(1)` to `while (hex_cmd < fw_end)`
- in `adm1266_check_ending`, changed `while(1)` to
`for (; !hex_cmd || hex_cmd >= fw_end;`
- reduce one level of indentation in `adm1266_nvmem_write`
with `continue` statement
- changed all `kzalloc` with `devm_kzalloc`
- initialize debugfs even when no firmware is present

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -392,14 +473,361 @@ static int adm1266_nvmem_read(void *priv, unsigned int offset, void *val,
+	memset(passwd_cmd, 0xFF, PMBUS_BLOCK_MAX);
+	passwd_cmd[ADM1266_PASSWD_CMD_LEN - 1] = ADM1266_UNLOCK_DEV;
+
+	/* password needs to be written twice correctly*/
```
nitpick: space after  `correctly */`

commodo replied:
same comment for other places

commodo commented:
1 minor nitpick;
will merge this tomorrow if no other objections
